src/parse_eqs.jl

```diff
@@ -0,0 +1,415 @@
+    if newfnbody.head == :(=)
+        newfnbody = Expr(:block, newfnbody)
+        debug && @show(newfnbody)
+    end
+
+    debug && (@show(newfnbody); println())
+
+    # Populate v_vars, v_assign, v_preamb
+    for (i, aa) in enumerate(newfnbody.args)
+        aavar = aa.args[1]
+        aaa = aa.args[2]
+        if isa(aaa, Expr)
+            fnexpr, def_fnexpr, auxfnexpr = _replacecalls!(aaa, aavar, v_vars)
+            push!(v_preamb, def_fnexpr.args[1] => def_fnexpr.args[2])
+            newfnbody.args[i] = fnexpr
+            if auxfnexpr.head != :nothing
+                push!(v_preamb, auxfnexpr.args[1] => auxfnexpr.args[2])
+            end
+            push!(v_vars, aavar)
+            #
+        elseif isa(aaa, Symbol) # occurs when there is a simple assignment
+            bb = subs(aa, Dict(aaa => :(identity($aaa))))
+            bbvar = bb.args[1]
+            bbb = bb.args[2]
+            fnexpr, def_fnexpr, auxfnexpr = _replacecalls!(bbb, bbvar, v_vars)
+            push!(v_preamb, def_fnexpr.args[1] => def_fnexpr.args[2])
+            newfnbody.args[i] = fnexpr
+            if auxfnexpr.head != :nothing
+                push!(v_preamb, auxfnexpr.args[1] => auxfnexpr.args[2])
+            end
+            push!(v_vars, bbvar)
+            #
+        elseif isa(aaa, Number)
+            push!(v_assign, (i, aa))
+            #
+        else #needed?
+            @show(aa.args[2], typeof(aa.args[2]))
+            error("Different from `Expr`, `Symbol` or `Number`")
+            #
+        end
+    end
+
+    debug && @show(v_vars, v_assign, v_preamb)
+
+    # Define premable (temporary allocations)
+    preamble = _newpreamble(v_vars, v_preamb, fnargs)
+
+    # Clean-up numeric assignments in preamble and body
+    for kk in reverse(v_assign)
+        newfnbody = subs(newfnbody, Dict(kk[2].args[1] => kk[2].args[2]))
+        preamble = subs(preamble, Dict(kk[2].args[1] => kk[2].args[2]))
+        deleteat!(newfnbody.args, kk[1])
+    end
+
+    # Check consistency of retvar
+    @assert(v_vars[end] == preamble.args[end].args[1])
+
+    # Guessed return variable (scalar equations); last included in v_vars/preamble
+    # retvar = preamble.args[end].args[1]
+    retvar = v_vars[end]
+
+    # Bring back indexed variables into place
+    preamble = subs( preamble, d_indx)
+    newfnbody = subs( newfnbody, d_indx)
+    retvar = subs( retvar, d_indx)
+
+    return preamble, newfnbody, retvar
+end
+
+
+"""
+`_rename_indexedvars(fnbody)`
+
+Renames the indexed variables (using `Espresso.genname()`) that
+exists in `fnbody`. Returns `fnbody` with the renamed variables,
+a vector of symbols with the new variables, and a dictionary
+that links the new variables to the old indexed variables.
+
+"""
+function _rename_indexedvars(fnbody)
+
+    v_indexed = Symbol[]
+    d_indx = Dict{Symbol, Expr}()
+
+    !(isindexed(fnbody)) && return fnbody, v_indexed, d_indx
+
+    # Obtain variables
+    vvars = get_vars(fnbody, rec=true)
```

Comment:
I found myself consistently forgetting to put `rec=true` so I made another function - `find_vars` - which is exactly like `get_vars`, but with `rec=true` by default. 

You may also be interested in `findex()` function which finds an expression matching a pattern. For example, here's how I would write `_rename_indexedvars`:

```
function _rename_indexedvars(fnbody)
    indexed_vars = findex(:(_X[_i...]), fnbody)
    st = Dict(ivar => genname() for ivar in indexed_vars)
    new_fnbody = subs(fnbody, st)
    return new_fnbody, collect(values(st)), Dict(v => k for (k, v) in st)
end
```

Reply:
Thanks for the comment. If you agree, I will incorporate this to the function.

Reply:
Quick question: Does `findex` allows to find cases like ` _x = _x + something`?

Reply:
> Thanks for the comment. If you agree, I will incorporate this to the function.

Sure!

> Quick question: Does findex allows to find cases like _x = _x + something?

Assuming `_x` is matched against the same thing on both sides, yes. All in all, `findex` simply traverses an expression and calls `matchex` to each node, so you can look at `@doc matchex` to learn about other options. 

Since you replace certain parts with the others in your code, you may also be interested in `rewrite()` function which takes an _input pattern_ and _input expression_ and tries to rewrite it according to an _output pattern_. Again, see `@doc rewrite` for details.

Reply:
Thanks for the suggestion, I'll check the docs of both, `matchex` and `rewrite`.